src/enzyme_ad/jax/Passes/EnzymeHLOOpt.cpp

```diff
@@ -304,11 +304,91 @@ class StaticSlice {
   }
 };
 
-struct NoopSlice final : OpRewritePattern<mlir::stablehlo::SliceOp> {
-  using OpRewritePattern::OpRewritePattern;
+LogicalResult failIfDynamicShape(Operation *op, PatternRewriter &rewriter) {
+  for (auto result : op->getResults()) {
+    auto type = dyn_cast<RankedTensorType>(result.getType());
+    if (!type || !type.hasStaticShape())
+      return rewriter.notifyMatchFailure(
+          op, "unsupported dynamic shape for output.");
+  }
+
+  for (auto operand : op->getOperands()) {
+    auto type = dyn_cast<RankedTensorType>(operand.getType());
+    if (!type || !type.hasStaticShape())
+      return rewriter.notifyMatchFailure(
+          op, "unsupported dynamic shape for input.");
+  }
+
+  return success();
+}
+
+LogicalResult failIfFuncOpInterfaceHasAttr(Operation *op, StringRef attrName,
+                                           PatternRewriter &rewriter) {
+  if (auto func = op->getParentOfType<FunctionOpInterface>()) {
+    if (func->hasAttrOfType<UnitAttr>(attrName))
+      return rewriter.notifyMatchFailure(op, "disabled by attribute.");
+  }
+
+  return success();
+}
+
+static constexpr StringRef kDisablePatternAttrName =
+    "enzymexla.disable_hlo_opts";
+
+template <typename OpTy, typename Child>
+struct CheckedOpRewritePattern : public OpRewritePattern<OpTy> {
+  using Base = OpRewritePattern<OpTy>;
+  using Base::Base;
+
+  LogicalResult matchAndRewrite(OpTy op,
+                                PatternRewriter &rewriter) const override {
+    LogicalResult res =
+        failIfFuncOpInterfaceHasAttr(op, kDisablePatternAttrName, rewriter);
+    if (res.failed())
+      return res;
+
+    if (!((Child *)this)->supportsDynamicShapes()) {
```

Comment:
I don't see much value in CRTP here. The classes already have a vtable and we get to patterns through a vitrual call. A couple more won't change much, and I don't think the pattern infra cost is dominated by virtual calls. Maintainability-wise, it will be easier without.

Also, it may be worth it to declare `matchAndRewrite` final.

Reply:
@wsmoses what do you think?

Reply:
I'm okay either way, slight pref to leave as CRTP for ease/perf [esp since we already have]